pkg/operator/ceph/csi/csidriver_object.go

```diff
@@ -0,0 +1,194 @@
+		csiDriver.ObjectMeta.ResourceVersion = driver.ObjectMeta.ResourceVersion
+		_, err = csidrivers.Update(ctx, csiDriver, metav1.UpdateOptions{})
+		if err == nil {
+			logger.Infof("CSIDriver object updated for driver %q", name)
+		}
+		return err
+	}
+
+	if apierrors.IsNotFound(err) {
+		_, err = csidrivers.Create(ctx, csiDriver, metav1.CreateOptions{})
+		if err == nil {
+			logger.Infof("CSIDriver object created for driver %q", name)
+		}
+	}
+
+	return err
+}
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func (v1csi v1CsiDriver) createCSIDriverInfo(ctx context.Context, clientset kubernetes.Interface, name, fsGroupPolicy string) error {
+	attach := true
+	mountInfo := false
+	// Create CSIDriver object
+	csiDriver := &v1k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1k8scsi.CSIDriverSpec{
+			AttachRequired: &attach,
+			PodInfoOnMount: &mountInfo,
+		},
+	}
+	if fsGroupPolicy != "" {
+		policy := v1k8scsi.FSGroupPolicy(fsGroupPolicy)
+		csiDriver.Spec.FSGroupPolicy = &policy
+	}
+	csidrivers := clientset.StorageV1().CSIDrivers()
+	driver, err := csidrivers.Get(ctx, name, metav1.GetOptions{})
+	if err == nil {
+		// As FSGroupPolicy field is immutable, should be set only during create time.
+		// if the request is to change the FSGroupPolicy, we are deleting the CSIDriver object and creating it.
+		if driver.Spec.FSGroupPolicy != nil && csiDriver.Spec.FSGroupPolicy != nil && *driver.Spec.FSGroupPolicy != *csiDriver.Spec.FSGroupPolicy {
+			return v1csi.reCreateCSIDriverInfo(ctx)
+		}
+
+		// For csidriver we need to provide the resourceVersion when updating the object.
+		// From the docs (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata)
+		// > "This value MUST be treated as opaque by clients and passed unmodified back to the server"
+		csiDriver.ObjectMeta.ResourceVersion = driver.ObjectMeta.ResourceVersion
+		_, err = csidrivers.Update(ctx, csiDriver, metav1.UpdateOptions{})
+		if err == nil {
+			logger.Infof("CSIDriver object updated for driver %q", name)
+		}
+		return err
+	}
+
+	if apierrors.IsNotFound(err) {
+		_, err = csidrivers.Create(ctx, csiDriver, metav1.CreateOptions{})
+		if err == nil {
+			logger.Infof("CSIDriver object created for driver %q", name)
+		}
+	}
+
+	return err
+}
+
+func (v1csi v1CsiDriver) reCreateCSIDriverInfo(ctx context.Context) error {
+	csiDriver := v1csi.csiDriver
+	err := v1csi.csiClient.Delete(ctx, csiDriver.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete CSIDriver object for driver %q", csiDriver)
+	}
+	logger.Infof("CSIDriver object deleted for driver %q", csiDriver)
+	_, err = v1csi.csiClient.Create(ctx, v1csi.csiDriver, metav1.CreateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to recreate CSIDriver object for driver %q", csiDriver)
```

Comment:
use `csiDriver.Delete` instead of `v1csi.csiClient.Delete` and same for create also